From ca18805bfbcc4b0b8a0d69ad032b029d1fe77720 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20Pale=C4=8Dek?= <palecek.pavel@gmail.com>
Date: Fri, 21 Sep 2018 19:27:32 +0200
Subject: [PATCH 1/3] tapio/live-server/#272 issue fix. It was caused by
 multiple inserts of refresh script (injected.html), but not reflected in
 Content-Length response header.

Bug shows up on ajax inlining svg files.
---
 index.js | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/index.js b/index.js
index 77a2a7e..407add0 100644
--- a/index.js
+++ b/index.js
@@ -79,12 +79,18 @@ function staticServer(root) {
 
 		function inject(stream) {
 			if (injectTag) {
-				// We need to modify the length given to browser
-				var len = INJECTED_CODE.length + res.getHeader('Content-Length');
-				res.setHeader('Content-Length', len);
 				var originalPipe = stream.pipe;
 				stream.pipe = function(resp) {
-					originalPipe.call(stream, es.replace(new RegExp(injectTag, "i"), INJECTED_CODE + injectTag)).pipe(resp);
+					originalPipe.call(stream, es.through(
+						function (data) {
+							var str = data.toString().replace(new RegExp(injectTag, "i"), INJECTED_CODE + injectTag)
+							// We need to modify the length given to browser
+							// before data emitted
+							res.setHeader('Content-Length', str.length);
+							this.emit('data', str)
+						}
+					))
+					.pipe(resp);
 				};
 			}
 		}

From eca7a6b3deaf2bb911f1514a82efd904e83422c9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20Pale=C4=8Dek?= <palecek.pavel@gmail.com>
Date: Fri, 21 Sep 2018 21:59:19 +0200
Subject: [PATCH 2/3] Revert "tapio/live-server/#272 issue fix. It was caused
 by multiple inserts of refresh script (injected.html), but not reflected in
 Content-Length response header."

This reverts commit ca18805bfbcc4b0b8a0d69ad032b029d1fe77720.
---
 index.js | 14 ++++----------
 1 file changed, 4 insertions(+), 10 deletions(-)

diff --git a/index.js b/index.js
index 407add0..77a2a7e 100644
--- a/index.js
+++ b/index.js
@@ -79,18 +79,12 @@ function staticServer(root) {
 
 		function inject(stream) {
 			if (injectTag) {
+				// We need to modify the length given to browser
+				var len = INJECTED_CODE.length + res.getHeader('Content-Length');
+				res.setHeader('Content-Length', len);
 				var originalPipe = stream.pipe;
 				stream.pipe = function(resp) {
-					originalPipe.call(stream, es.through(
-						function (data) {
-							var str = data.toString().replace(new RegExp(injectTag, "i"), INJECTED_CODE + injectTag)
-							// We need to modify the length given to browser
-							// before data emitted
-							res.setHeader('Content-Length', str.length);
-							this.emit('data', str)
-						}
-					))
-					.pipe(resp);
+					originalPipe.call(stream, es.replace(new RegExp(injectTag, "i"), INJECTED_CODE + injectTag)).pipe(resp);
 				};
 			}
 		}

From 68050b30201cec270f284e4b3d23841c34821076 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20Pale=C4=8Dek?= <palecek.pavel@gmail.com>
Date: Fri, 21 Sep 2018 22:00:43 +0200
Subject: [PATCH 3/3] tapio/live-server#272 issue fix. It was caused by
 multiple inserts of refresh script (injected.html), but not reflected in
 Content-Length response header.

Bug shows up on ajax inlining svg files.
---
 index.js | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/index.js b/index.js
index 77a2a7e..3f84444 100644
--- a/index.js
+++ b/index.js
@@ -42,8 +42,9 @@ function staticServer(root) {
 		if (req.method !== "GET" && req.method !== "HEAD") return next();
 		var reqpath = isFile ? "" : url.parse(req.url).pathname;
 		var hasNoOrigin = !req.headers.origin;
-		var injectCandidates = [ new RegExp("</body>", "i"), new RegExp("</svg>"), new RegExp("</head>", "i")];
+		var injectCandidates = [ new RegExp("</body>", "i"), new RegExp("</svg>", "g"), new RegExp("</head>", "i")];
 		var injectTag = null;
+		var injectCount = 0
 
 		function directory() {
 			var pathname = url.parse(req.originalUrl).pathname;
@@ -53,15 +54,16 @@ function staticServer(root) {
 		}
 
 		function file(filepath /*, stat*/) {
-			var x = path.extname(filepath).toLocaleLowerCase(), match,
+			var x = path.extname(filepath).toLocaleLowerCase(), matches,
 					possibleExtensions = [ "", ".html", ".htm", ".xhtml", ".php", ".svg" ];
 			if (hasNoOrigin && (possibleExtensions.indexOf(x) > -1)) {
 				// TODO: Sync file read here is not nice, but we need to determine if the html should be injected or not
 				var contents = fs.readFileSync(filepath, "utf8");
 				for (var i = 0; i < injectCandidates.length; ++i) {
-					match = injectCandidates[i].exec(contents);
-					if (match) {
-						injectTag = match[0];
+					matches = contents.match(injectCandidates[i]);
+					injectCount = matches && matches.length || 0
+					if (injectCount) {
+						injectTag = matches[0];
 						break;
 					}
 				}
@@ -80,7 +82,7 @@ function staticServer(root) {
 		function inject(stream) {
 			if (injectTag) {
 				// We need to modify the length given to browser
-				var len = INJECTED_CODE.length + res.getHeader('Content-Length');
+				var len = INJECTED_CODE.length * injectCount + res.getHeader('Content-Length');
 				res.setHeader('Content-Length', len);
 				var originalPipe = stream.pipe;
 				stream.pipe = function(resp) {
